x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ApplicationPrivilege.java

```diff
@@ -0,0 +1,275 @@
+        PARSER.declareString(ConstructingObjectParser.constructorArg(), Fields.NAME);
+        PARSER.declareStringArray(ConstructingObjectParser.constructorArg(), Fields.ACTIONS);
+        PARSER.declareObject(ConstructingObjectParser.optionalConstructorArg(), (parser, context) -> parser.map(), Fields.METADATA);
+        PARSER.declareField((priv, fieldValue) -> {
+            if (DOC_TYPE_VALUE.equals(fieldValue) == false) {
+                throw new IllegalStateException("XContent has wrong " + Fields.TYPE.getPreferredName() + " field " + fieldValue);
+            }
+        }, (parser, acceptType) -> {
+            if (acceptType == false) {
+                throw new IllegalStateException("Field " + Fields.TYPE.getPreferredName() + " cannot be specified here");
+            }
+            return parser.text();
+        }, Fields.TYPE, ObjectParser.ValueType.STRING);
+    }
+
+    public static final Pattern VALID_APPLICATION = Pattern.compile("^[a-z][A-Za-z0-9_-]{2,}$");
+    public static final Pattern VALID_NAME = Pattern.compile("^[a-z][a-zA-Z0-9_.-]*$");
+
+    public static final Function<String, ApplicationPrivilege> NONE = app -> new ApplicationPrivilege(app, "none", new String[0]);
+
+    private final String application;
+    private final String[] patterns;
+    private final Map<String, Object> metadata;
+
+    public ApplicationPrivilege(String application, String privilegeName, Collection<String> patterns, Map<String, Object> metadata) {
+        this(application, Collections.singleton(privilegeName), patterns.toArray(new String[patterns.size()]), metadata, true);
+    }
+
+    public ApplicationPrivilege(String application, String privilegeName, String... patterns) {
+        this(application, Collections.singleton(privilegeName), patterns, emptyMap(), true);
+    }
+
+    private ApplicationPrivilege(String application, Set<String> name, String[] patterns, Map<String, Object> metadata,
+                                 boolean validateNames) {
+        super(name, patterns);
+        this.application = application;
+        this.patterns = patterns;
+        this.metadata = new HashMap<>(metadata == null ? emptyMap() : metadata);
+        validate(validateNames);
+    }
+
+    public String getApplication() {
+        return application;
+    }
+
+    /**
+     * If this privilege has a single name, returns that name. Otherwise throws {@link IllegalStateException}.
+     * @see #name()
+     */
+    public String getPrivilegeName() {
+        if (name.size() == 1) {
+            return name.iterator().next();
+        } else {
+            throw new IllegalStateException(this + " has a multivariate name: " + Strings.collectionToCommaDelimitedString(name));
+        }
+    }
+
+    public Map<String, Object> getMetadata() {
+        return Collections.unmodifiableMap(metadata);
+    }
+
+    private void validate(boolean validateNames) {
+        if (Regex.isSimpleMatchPattern(application) == false) {
```

Comment:
We should. It all got a bit ugly somewhere along the way.
 
This exists, purely so `superuser` can have _all_ privileges on _all_ applications.
`SUPERUSER_ROLE_DESCRIPTOR` has `application("*").privileges("*").resources("*")`

Checks in `PutPrivilegesRequest` and `HasPrivilegesRequest` prevent applications wildcards being used in privilege definitions, but they are possible in roles.

The problem really comes from using a single class for defining privileges themselves, and for representing the privileges that a role might grant. You can't define a privilege with a wildcard application, but you can grant one.

I've put in a fix for this check, but the whole thing still feels ugly.

Reply:
> The problem really comes from using a single class for defining privileges themselves, and for representing the privileges that a role might grant. You can't define a privilege with a wildcard application, but you can grant one.

Based on your response I don't feel like you're happy with how this is implemented. How do you feel about us continuing to review this PR and instead of merging to master we start a feature branch where these classes can be separated?